Move ListOptions into meta

smarterclayton opened the pull request:
metav1.ListOptions is moving to k8s.io/apimachinery/pkg/apis/meta/v1.  The internal version will be reserved for server use (clients will use metav1), and all references changed.

Also, all references to the Namespace* constants are being moved to point to metav1.

smarterclayton committed:
Create meta/internalversion for ListOptions

Move over only the conversions that are needed, create a new scheme that
is private to meta and only accessible via ParameterCodec. Move half of
pkg/util/labels/.readonly to pkg/apis/meta/v1/labels.go

smarterclayton committed:
Make all groups use metav1.AddToGroupVersion for ListOptions

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40178)
<!-- Reviewable:end -->

smarterclayton committed:
Switch genericapiserver to use metainternalversion.ListOptions

Decouple ListOption parsing from the scheme - instead, it is a property
of the server (and clients should use metav1.ListOptions for now).

smarterclayton committed:
Update registries to use metainternalversion list options

smarterclayton committed:
Tests that check metav1

smarterclayton committed:
Convert core code to metav1.ListOptions

smarterclayton committed:
Remove federation api defaulters

smarterclayton committed:
Generators should reference metav1.ListOptions (informer,client,fake)

smarterclayton committed:
refactor: move ListOptions references to metav1

smarterclayton committed:
Break dependency on testapi from core packages

smarterclayton committed:
generated: informer,client

smarterclayton committed:
Change the string serialization of the Nothing selector

The previous option was not serializable, and was used in questionable
scenarios to represent what we also use Everything() for. The individual
context needs to make the decision about everything vs nothing (and
nothing has no real concept in our API today) so this seems safe.

smarterclayton commented:
That was worse than I thought.  I'll need to bump gengo to fix deepcopy gen (which is the DNM commit in here), and I had to change the `String()` value for the `labels.Nothing()` helper (the commit message has more info).

The vast majority of code simply uses metav1.  The server uses metainternalversion with conversions.

deads2k commented:
Definitely on my Monday list.  I'm looking forward to this.

sttts commented on staging/src/k8s.io/client-go/pkg/apis/policy/v1alpha1/register.go:
```diff
@@ -0,0 +1,55 @@
+/*
```
Doesn't belong into this commit?

smarterclayton replied:
I did both of them together because staging/copy.sh is / was broken for a bit.  And both need to change at once, so I wanted to be clear that both sides are the same.

sttts commented on pkg/genericapiserver/endpoints/apiserver_test.go:
```diff
@@ -117,20 +118,11 @@ func newMapper() *meta.DefaultRESTMapper {
-		Watch           bool   `json:"watch,omitempty"`
-		ResourceVersion string `json:"resourceVersion,omitempty"`
-		TimeoutSeconds  *int64 `json:"timeoutSeconds,omitempty"`
-	}
```
👍

deads2k commented on vendor/k8s.io/gengo/examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -184,10 +184,18 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
 
 		if pkgNeedsGeneration {
 			glog.V(3).Infof("Package %q needs generation", i)
+			path := pkg.Path
```
Can you give an example before and after for this chunk when you open the upstream.  It's not obvious at a glance what this is doing and I'm wondering if you already added similar code elsewhere.

smarterclayton replied:
Yeah, basically your "CanonicalizeImports" strips the information that gengo needs in order to output to the correct location.  I think there is a better fix in gengo, but I don't have the brainpower to refine it yet (and haven't seen all the cases).

Basically "k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime" is a valid go path (to various tools), and also to the output base location that we need to write to.

deads2k commented on pkg/client/cache/listwatch_test.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package cache
+package cache_test
```
@sttts I suspect we're going to end up with a `pkg/client/testing` like we did for `apimachinery`.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/register.go:
```diff
@@ -0,0 +1,77 @@
+const GroupName = "meta.k8s.io"
+
+// Scheme is the registry for any type that adheres to the meta API spec.
+var scheme = runtime.NewScheme()
```
Is this an intermediate step or do you anticipate parameters having their own scheme?  If so, can I get an updated comment and a better name when you rebase?

smarterclayton replied:
I think meta will have a scheme for the generic clients and generic server code - basically the server *stops* looking at the normal scheme for standard types (things that *must* behave a certain way in meta/v1 compatible servers).  That would include other things besides parameters.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/register.go:
```diff
@@ -42,10 +42,23 @@ func AddToGroupVersion(scheme *runtime.Scheme, groupVersion schema.GroupVersion)
 }
+
+// scheme is the registry for the common types that adhere to the meta v1 API spec.
+var scheme = runtime.NewScheme()
```
again, parameter scheme?  Will I later be adding deleteoptions here?  If so, better comment, better name.

smarterclayton replied:
We will eventually be adding things that are not parameters, things like the structured responses for get / describe and probably List.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -219,6 +219,17 @@ type ObjectMeta struct {
+	// NamespaceNone is the argument for a context when there is no namespace.
+	NamespaceNone string = ""
+	// NamespaceSystem is the system namespace where we place system components.
+	NamespaceSystem string = "kube-system"
```
This one looks off.  Why does kube-system move?

smarterclayton replied:
Mostly to break dependencies on pkg/api.  The kube-system namespace is no more or less special than "default" in that it has meaning for mirror pods.

sttts replied:
This should go to either client-go or stay in kube. It's not generic and only used in kube components. k8s.io/apiserver doesn't need it.

smarterclayton replied:
I don't now that I agree.  The convention is that any Kube like API that knows about namespaces knows what `kube-system` is.  How about this, we argue more about where it should end up and then we move it separately (I'd rather not unwind it here).

sttts replied:
No reason to stop this PR, sure. What can be assumed about`kube-system` under this convention? In a generic `k8s.io/apiserver` world we don't have any controller, no "kubernetes" service, no addons, not even namespaces as resources.

deads2k commented on pkg/genericapiserver/endpoints/handlers/rest.go:
```diff
@@ -87,6 +88,8 @@ type RequestScope struct {
 	Kind        schema.GroupVersionKind
 	Subresource string
+
+	MetaGroupVersion schema.GroupVersion
```
This more precisely a parameter groupversion, correct?

deads2k commented on pkg/genericapiserver/server/genericapiserver.go:
```diff
@@ -59,6 +59,10 @@ type APIGroupInfo struct {
+	// MetaGroupVersion defaults to "meta.k8s.io/v1" and is the scheme group version used to decode
+	// common API implementations like ListOptions. Future changes will allow this to vary by group
+	// version (for when the inevitable meta/v2 group emerges).
+	MetaGroupVersion *schema.GroupVersion
```
ParameterGroupVersion?

smarterclayton replied:
I *think* this would be used for other meta related behaviors.  Basically it's asserting the server group version (which was the old name) which influences default behavior.  And should be reported in swagger.

smarterclayton replied:
So `ServerGroupVersion` is most accurate, but now that we're using `meta` for the group I slightly prefer `MetaGroupVersion`.

deads2k commented:
Relatively minor comments from me. Given importance, its blocking nature, and rebase difficulty, nothing I'd block a merge on.

smarterclayton committed:
generated: staging client-go

smarterclayton committed:
Test changes due to client-go staging changes

smarterclayton commented:
Created the gengo PR at kubernetes/gengo#37

deads2k commented:
lgtm.  I think we'll have a few more minor shuffles and maybe some renames, but this breaks the back of the problem.

deads2k commented:
bumping priority to avoid the continual rebases we're seeing already.

sttts commented:
@k8s-bot cvm gke e2e test this

smarterclayton committed:
bump(k8s.io/gengo):c118aa8edfff53fe5b69127a970f54b6cf3a7563

smarterclayton commented:
Rebased with vendored gengo, hopefully no issues.

smarterclayton commented:
So close

smarterclayton commented:
Govet failure, reapplying label

smarterclayton commented:
/approve

deads2k commented:
And the queue isn't taking it.  Awesome
